### PR TITLE
Rewrite CMake builds `install`/`distrib` targets.

### PR DESCRIPTION
**Pull request by kernhanda**
This change adds support for a proper `install` target for CMake builds. This allows for the usage of CPack to create packages for various systems and platforms. Once Halide is ready to do these things, the settings can be adjusted accordingly. For now, the focus is on creating tgz and zip packages for POSIX systems and Windows, respectively. This is done using CPack.

The `distrib` target that was previously available has been rewritten in terms of the `install` target and taking advantage of being able to call CPack through CMake. As was done previously, a Halide.tgz package is created for POSIX systems while a Halide.zip package is created for Windows systems.

`halide.cmake` also was fixed up to be more platform agnostic, specifically around library filenames and the presence of implibs for shared libraries on Windows.

For Windows builds specifically, the following changes were made:
* Enable support for no RTTI builds
* Enable support for OpenMP builds
* Enable building of all but one app in the `apps/` folder
* Fix emitted code for the C codegen to not result in an "inconsistent dll linkage" warning

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -375,149 +393,74 @@ else()
...
-foreach(EXT png)
-  set(DIR "${CMAKE_SOURCE_DIR}/tutorial/images")
-  file(GLOB FILES "${DIR}/*.${EXT}")
-  foreach(F ${FILES})
-    add_to_distrib(${F} tutorial/images)
-  endforeach()
-endforeach()
+# install
+
+install(FILES
```
Comment (abadams) [inline CMakeLists.txt]: How does HalideRuntime*.h get included?

**Comment by abadams** [comment]
Thank you very much for this. I just have one question about how the appropriate headers are found.

File: CMakeLists.txt (hunk 1 above)
Comment (kernhanda) [reply CMakeLists.txt]: Thanks for catching this. It was an unintentional omission. It's fixed now.

**Comment by kernhanda** [comment]
The issue in question has been fixed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,19 @@
...
+set(CPACK_PACKAGE_VENDOR "Halide")
+set(CPACK_RESOURCE_FILE_LICENSE "${CMAKE_SOURCE_DIR}/LICENSE.txt")
+set(CPACK_MONOLITHIC_INSTALL OFF)
+if (WIN32)
+  set(CPACK_GENERATOR "ZIP")
+else()
+  set(CPACK_GENERATOR "TGZ")
+endif()
+# Remove this to get package names that are formatted as
+# ${CPACK_PACKAGE_NAME}-${CPACK_PACKAGE_VERSION}-${CPACK_SYSTEM_NAME}.
```
Comment (steven-johnson) [inline CMakeLists.txt]: I don't quite understand the difference between these options; why would one want the commented-out variant?

File: CMakeLists.txt [hunk 3]
```diff
@@ -2,6 +2,19 @@
 project(Halide)
 cmake_minimum_required(VERSION 3.2)
```
Comment (steven-johnson) [inline CMakeLists.txt]: Is 3.2 the minimum for CPack?

File: CMakeLists.txt [hunk 4]
```diff
@@ -45,6 +58,11 @@ set(CMAKE_MACOSX_RPATH ON)
 # Export all symbols
 SET(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
 
+# # Allow for the distribution directory to be explicitly specified
```
Comment (steven-johnson) [inline CMakeLists.txt]: Why is this code commented out? (We generally avoid checking in commented-out code unless it's an example of what not to do or debugging aid)

File: CMakeLists.txt [hunk 5]
```diff
@@ -155,6 +173,9 @@ endif()
 function(halide_project name folder)
   add_executable("${name}" ${ARGN})
   if (MSVC)
+    if(NOT HALIDE_ENABLE_RTTI)
```
Comment (steven-johnson) [inline CMakeLists.txt]: It's a little surprising that CMake doesn't offer a standard option for 'enable/disable rtti'. (Maybe it does and I haven't checked?)

File: CMakeLists.txt [hunk 6]
```diff
@@ -333,15 +354,12 @@ endif()
 
 option(WITH_APPS "Build apps" ON)
 if (WITH_APPS)
-  if (NOT WIN32)
-    message(STATUS "Building apps enabled")
-    add_subdirectory(apps)
-  else()
-    message(WARNING "Apps cannot be built under Windows using cmake. Try the makefile in an msys2 shell.")
-  endif()
+  message(STATUS "Building apps enabled")
```
Comment (steven-johnson) [inline CMakeLists.txt]: Thank you so much for doing this! It's been on my neglected list for way too long.

File: Makefile [hunk 7]
```diff
@@ -1718,7 +1718,7 @@ install_qc: install $(HEXAGON_RUNTIME_LIBS)
 # we include in our distribution.
 $(BUILD_DIR)/halide_config.%: $(ROOT_DIR)/tools/halide_config.%.tpl
 	@mkdir -p $(@D)
-	cat $< | sed -e 's/$${LLVM_SYSTEM_LIBS}/${LLVM_SYSTEM_LIBS}/g' > $@
+	cat $< | sed -e 's/@HALIDE_SYSTEM_LIBS_RAW@/${LLVM_SYSTEM_LIBS}/g' > $@
```
Comment (steven-johnson) [inline Makefile]: Why is this HALIDE_SYSTEM_LIBS_RAW instead of HALIDE_SYSTEM_LIBS?

Comment (kernhanda) [reply Makefile]: `HALIDE_SYSTEM_LIBS_RAW` is the new value that is to be replaced in `tools/halide_config.bzl.tpl` and `tools/halide_config.cmake.tpl`. In CMake, the value will be automatically replaced by the value of the variable (within the CMake configuration), but with Makefiles, we have more flexibility through `sed`.

File: CMakeLists.txt (hunk 6 above)
Comment (kernhanda) [reply CMakeLists.txt]: NP!

File: CMakeLists.txt (hunk 5 above)
Comment (kernhanda) [reply CMakeLists.txt]: It does not. I think CMake tries to have just enough built-ins to get by, which is fair considering the number of compilers and platforms they support.

**Comment by steven-johnson** [comment]
LGTM pending the win32 issues.

File: CMakeLists.txt (hunk 3 above)
Comment (kernhanda) [reply CMakeLists.txt]: The minimum for the CPack module is at least 3.0.0, if not earlier, considering it is part of the CMake suite.

File: CMakeLists.txt (hunk 2 above)
Comment (kernhanda) [reply CMakeLists.txt]: The commented out variant starts being useful when there's automation in place to create packages from release branches/tags.

`Halide-v1.0.0-linux.tgz` and `Halide-v1.0.0-macos.tgz` vs. just one `Halide.tgz`.

File: CMakeLists.txt (hunk 4 above)
Comment (kernhanda) [reply CMakeLists.txt]: Fixed by removing the block. This was no longer needed once everything got moved to use `install` and CPack for the `distrib` target.

**Comment by kernhanda** [comment]
Looks like the travis-ci/push build failed due to infra issues. Any way it can be restarted?

**Comment by kernhanda** [comment]
Disabled `apps/camera_pipe` from building on windows, since that seems to suffer from the same issue.

**Comment by steven-johnson** [comment]
Since this looks imminent, I'd like to wait for this to land before doing a release.

**Pull request merged**